---
title: "Total Variant Count Analysis"
author: "Nate Olson"
date: '`r Sys.Date()`'
output: 
    bookdown::html_document2:
        toc: true
        toc_float: true
        df_print: paged
        code_folding: hide
---

```{r load_packages, message = FALSE, echo = FALSE}
library(happyR)
library(tidyverse)
```

# Objective
Get variant count estimates for CCS variant callsets inside and outside the GIAB v3.3.2 benchmark regions. 

# Method Evaluation
Total QUERY.TRUTH and ??? will be used for the number of variants inside and outside the benchmark regions respecitively. 
Accurate variant counts is challenging due to different variant representions in VCF files. 
To ensure that the variants are appropriately represented for callset comparison, total counts were compared between benchmarking results using with and without partial credit. 

## Loading Data for Method Evalaution
```{r}
## Loading data
hap_list <- list(
    GATK4_no_partial = "data/happy_output/results/result_1",
    DeepVar_no_partial = "data/happy_output/results/result_2",
    GATK4_w_partial = "data/happy_output_gatk4DeepVar_wpartial/results/result_1",
    DeepVar_w_partial = "data/happy_output_gatk4DeepVar_wpartial/results/result_2") %>%
    map(read_happy)

## Creating a tidy data frame
extend_df <- hap_list %>% 
    map("extended") %>% 
    bind_rows(.id = "query_method")

ext_trim_df <- extend_df %>% 
    ## Excluding non-HG002 genome specific stratifications
    filter(!str_detect(Subset, "HG00[1,3,4,5]")) %>% 
    ## excluding strat with < 1000 obs - lack of statistical power
    filter(Subset.Size > 1000) %>% 
    ## Subset must have at least one variant in region
    filter(Subset.IS_CONF.Size > 0)
```

Total count comparisons
```{r}
ext_trim_df %>% filter(Subtype == "*", Subset == "*", Filter == "PASS") %>% 
    arrange(Type, query_method) %>% 
    select(query_method, Type, QUERY.TOTAL, QUERY.UNK) %>% 
    gather("count_type", "count", - query_method, -Type) %>% 
    mutate(partial = if_else(str_detect(query_method, "no"), "no", "with"),
           method = str_remove(query_method, "_.*")) %>% 
    ggplot() + geom_col(aes(x = method, y = count, fill = partial), position = "dodge") +
    facet_wrap(count_type~Type, scales = "free") + theme_bw()
```

```{r}
ext_trim_df %>% filter(Subtype == "*", Subset == "*", Filter == "PASS") %>% 
    arrange(Type, query_method)  %>% 
    select(query_method, Type, QUERY.TOTAL, QUERY.UNK) 
```

```{r}
ext_trim_df %>% filter(Subtype == "*", Subset == "*", Filter == "PASS") %>% 
    arrange(Type, query_method)  %>% 
    mutate(bias_rate = (QUERY.TP - TRUTH.TP)/ TRUTH.TP,
           Inside = QUERY.TOTAL - QUERY.UNK,
           Outside = QUERY.UNK) %>% 
    select(query_method, Type, Inside, Outside, bias_rate)
```

